### PR TITLE
implement SDK telemetry for ASP.NET 5 applications

### PR DESCRIPTION
**Pull request by SergKanz**

### CODE REVIEW

File: src/Microsoft.ApplicationInsights.AspNet/SdkVersionUtils.cs [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+    using System.Reflection;
+
+    internal class SdkVersionUtils
+    {
+        internal const string VersionPrefix = "aspnet5: ";
+
+        internal static string GetAssemblyVersion()
+        {
+            return typeof(SdkVersionUtils).GetTypeInfo().Assembly.GetCustomAttributes<AssemblyFileVersionAttribute>()
+                    .First()
```
Comment (abaranch) [inline src/Microsoft.ApplicationInsights.AspNet/SdkVersionUtils.cs]: Is it always available?

**Comment by abaranch** [comment]
Approved

File: src/Microsoft.ApplicationInsights.AspNet/SdkVersionUtils.cs [hunk 2]
```diff
@@ -0,0 +1,17 @@
+﻿namespace Microsoft.ApplicationInsights.AspNet
+{ 
+    using System.Linq;
+    using System.Reflection;
+
+    internal class SdkVersionUtils
```
Comment (olegsych) [inline src/Microsoft.ApplicationInsights.AspNet/SdkVersionUtils.cs]: Can we convert this class to an `IContextInitializer` or an `ITelemetryInitializer`?

File: src/Microsoft.ApplicationInsights.AspNet/SdkVersionUtils.cs (hunk 1 above)
Comment (olegsych) [inline src/Microsoft.ApplicationInsights.AspNet/SdkVersionUtils.cs]: (Reflection + Linq) x 2 middlewares per request. :scream: Can we cache this value?

File: test/Microsoft.ApplicationInsights.AspNet.Tests/RequestTrackingMiddlewareTest.cs [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+            };
+            var middleware = new RequestTrackingMiddleware(nextMiddleware, MockTelemetryClient());
+
+            await middleware.Invoke(new DefaultHttpContext(), new RequestTelemetry());
+
+            Assert.NotEmpty(sentTelemetry.Context.GetInternalContext().SdkVersion);
+            Assert.Contains("aspnetv5", sentTelemetry.Context.GetInternalContext().SdkVersion);
+        }
+
+        private TelemetryClient MockTelemetryClient()
```
Comment (olegsych) [inline test/Microsoft.ApplicationInsights.AspNet.Tests/RequestTrackingMiddlewareTest.cs]: This looks similar to the helper method in the `ExceptionTrackingMiddlewareTest`. Can we share it?

File: src/Microsoft.ApplicationInsights.AspNet/ExceptionTrackingMiddleware.cs [hunk 4]
```diff
@@ -13,11 +14,13 @@ public sealed class ExceptionTrackingMiddleware
     {
         private readonly RequestDelegate next;
         private readonly TelemetryClient telemetryClient;
+        private readonly string sdkVersion;
 
         public ExceptionTrackingMiddleware(RequestDelegate next, TelemetryClient client)
         {
             this.next = next;
             this.telemetryClient = client;
+            this.sdkVersion = SdkVersionUtils.VersionPrefix + SdkVersionUtils.GetAssemblyVersion();
```
Comment (olegsych) [inline src/Microsoft.ApplicationInsights.AspNet/ExceptionTrackingMiddleware.cs]: Why are the middleware components responsible for setting the SdkVersion?

File: src/Microsoft.ApplicationInsights.AspNet/RequestTrackingMiddleware.cs [hunk 5]
```diff
@@ -16,14 +19,16 @@ public sealed class RequestTrackingMiddleware
...
         
         public RequestTrackingMiddleware(RequestDelegate next, TelemetryClient client)
         {
             this.telemetryClient = client;
             this.next = next;
+            this.sdkVersion = SdkVersionUtils.VersionPrefix + SdkVersionUtils.GetAssemblyVersion();
         }
 
-        public async Task Invoke(HttpContext httpContext)
+        public async Task Invoke(HttpContext httpContext, RequestTelemetry telemetry)
```
Comment (olegsych) [inline src/Microsoft.ApplicationInsights.AspNet/RequestTrackingMiddleware.cs]: :+1:

File: src/Microsoft.ApplicationInsights.AspNet/SdkVersionUtils.cs (hunk 1 above)
Comment (SergKanz) [inline src/Microsoft.ApplicationInsights.AspNet/SdkVersionUtils.cs]: We are using the same code in Core

File: src/Microsoft.ApplicationInsights.AspNet/SdkVersionUtils.cs (hunk 2 above)
Comment (SergKanz) [inline src/Microsoft.ApplicationInsights.AspNet/SdkVersionUtils.cs]: The idea here is only mark auto-generated data with aspnet5 keyword, not everything

File: test/Microsoft.ApplicationInsights.AspNet.Tests/RequestTrackingMiddlewareTest.cs (hunk 3 above)
Comment (SergKanz) [inline test/Microsoft.ApplicationInsights.AspNet.Tests/RequestTrackingMiddlewareTest.cs]: makes sense. I guess we will have more usage of this helper method

File: src/Microsoft.ApplicationInsights.AspNet/SdkVersionUtils.cs (hunk 2 above)
Comment (olegsych) [inline src/Microsoft.ApplicationInsights.AspNet/SdkVersionUtils.cs]: Ah, I remember now. It's the approach neither of us liked :smile:. Why are we doing it again? Does this require changes in our telemetry reports?

File: src/Microsoft.ApplicationInsights.AspNet/ExceptionTrackingMiddleware.cs (hunk 4 above)
Comment (SergKanz) [inline src/Microsoft.ApplicationInsights.AspNet/ExceptionTrackingMiddleware.cs]: The idea here is to count how many people are using auto-collection, not making TrackException manually

File: src/Microsoft.ApplicationInsights.AspNet/SdkVersionUtils.cs (hunk 2 above)
Comment (SergKanz) [inline src/Microsoft.ApplicationInsights.AspNet/SdkVersionUtils.cs]: I just believe in this pattern. What can I do to convince that it is good? Imagine - when you doing TrackMetric you can distinct whether it is call to TrackMetric or PerformanceCountersCollection module or @lukekim 's metric aggregation module. Will it be cool?

Comment (olegsych) [inline src/Microsoft.ApplicationInsights.AspNet/SdkVersionUtils.cs]: I understand what we are trying to achieve here (thanks for the reminder :smile:). However, putting this information in the SdkVersion field violates the [single responsibility principle](http://www.butunclebob.com/ArticleS.UncleBob.PrinciplesOfOod) of OO and the [first normal form](http://en.wikipedia.org/wiki/First_normal_form) of Database design, which leads to known consequences. Will the data pipeline separate the "source" information into a different field in storage schema for filtering and segmentation? Are we working with them to introduce a new field for this information in the data collection schema?

**Pull request merged**